### PR TITLE
feat: add 'article_view' reader activity data

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Implements `article_view` reader activity to the reader data library integrated with memberships and metering.

Content locked from memberships and front-end or back-end metering strategies should not push an `article_view` activity.

This PR also implements the `getUniqueActivitiesBy( name, iteratee )` helper function to allow fetching unique activity based on its data, e.g.:

```js
const uniquePostsViews = ras.getUniqueActivitiesBy( 'article_view', 'post_id' );
```

### How to test the changes in this Pull Request:

1. Make sure you have WC Memberships configured with content gating and metering (anonymous and registered readers)
2. On a fresh anonymous session visit an article
3. Open DevTools and navigate to the localStorage
4. Confirm there's a new item in the `newspack_reader_1_activity` array similar to this:
```js
{
  "action": "article_view",
  "data": {
    "post_id": 120,
    "permalink": "https://example.com/montes-egestas-a-neque-morbi-cras-maecenas-senectus-molestie-in-congue-consequat-sed/",
    "categories": [ 4, 3 ],
    "tags": [],
    "author": "johndoe"
  },
  "timestamp": 1687530871161
}
```
5. Navigate until content is locked and confirm the anonymous metering locked content does not push a new activity
6. Register and do the same with registered metering, navigate confirming activities are pushed until it locks and no longer pushes
7. Become a member and confirm activity pushes as a member
8. Deactivate WC Memberships, navigate and confirm activity continues to push correctly

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by leogermani — APPROVED** [review]
Works well after your last fix!

**Comment by miguelpeixe** [comment]
@leogermani @dkoo could use some help figuring out why the PHP linter is erroring here...

**Comment by leogermani** [comment]
I'm able to reproduce the lint error on my local... probably some conflict between package and php versions.. I'm looking into it

**Comment by miguelpeixe** [comment]
Thanks! I tried rebasing with master a few times without luck. It's definitely triggered by something in this code 😕

File: includes/plugins/wc-memberships/class-metering.php [hunk 1]
```diff
@@ -270,5 +279,18 @@ public static function is_logged_in_metering_allowed( $post_id = null ) {
...
+
+	/**
+	 * Store the article view activity push for use in the frontend metering
+	 * strategy.
+	 *
+	 * @param array $activity Activity data.
+	 *
+	 * @return array
+	 */
+	public static function get_article_view( $activity )  {
```
Comment (leogermani) [inline includes/plugins/wc-memberships/class-metering.php:291]: @miguelpeixe the linting error is here. two spaces after the parenthesis. Now I'm trying to understand why phpcs is breaking

Comment (miguelpeixe) [reply includes/plugins/wc-memberships/class-metering.php:291]: Ah ha! 🙇

**Comment by miguelpeixe** [comment]
While investigating a bug I found with metering having some inconsistent dispatches I ended up tweaking some code placement as well.

de00bd9b50e9208b009f1b48cd2322bb8442f49f moves the code to the frontend by localizing the activity push setup. It also corrects a bug with metering dispatch that I couldn't reproduce anywhere else. For some reason when it uses the array push strategy it doesn't run the dispatch, only when it uses the replaced push handler, it becomes a coin toss. I guess it might have to do with the method already running inside a push... Moving to calling the dispatch directly from RAS fixes the issue and is also a better practice since the code is already in a push.

**Comment by miguelpeixe** [comment]
If you could take another look @leogermani . Sorry for the back and forth 😬

**Comment by leogermani** [comment]
I'll have a look. 

In the meantime, this is the reason for the WPCS to crash: https://github.com/WordPress/WordPress-Coding-Standards/commit/7cd46bed1e6a7a2af3fe24c7f4a044da3076d8f4

TL;DR is:
* If you run phpcs on PHP 8, it will crash on some sniffs.
* recommendation is to run phpcs on php 7.4
* There is a [tag](https://github.com/WordPress/WordPress-Coding-Standards/releases/tag/2.3.0-fix-vsprintf) with the fixes we could use, but those also include other unstable changes...

I guess we should look to see if it's possible to configure circleCI to use php 7.4 for this particular job. If not, use the [unstable tag](https://github.com/WordPress/WordPress-Coding-Standards/releases/tag/2.3.0-fix-vsprintf).

**Comment by miguelpeixe** [comment]
Thank you!

Nice finding on the linter issue, should we open a gh issue for dedicated discussion?

**Review by leogermani — APPROVED** [review]
I like this approach much better!
